### PR TITLE
Restructured argument list in "[" and "[<-" methods

### PR DESCRIPTION
**Pull request by tscheypidi**
Restructured argument list in "[" and "[<-" methods to make it compatible with current R-develop. In addition this fixes also #150

### CODE REVIEW

File: R/magpie-class.R [hunk 1]
```diff
@@ -270,9 +270,12 @@ setClass("magpie", contains = "array", prototype = array(0, c(0, 0, 0)))
 }
 
 #' @exportMethod [
-setMethod("[", #nolint
+setMethod("[", # nolint
   signature(x = "magpie"),
-  function(x, i, j, k, drop = FALSE, pmatch = FALSE, invert = FALSE, dim = NULL) {
+  function(x, i, j, k, ..., pmatch = FALSE, invert = FALSE, dim = NULL, drop = FALSE) {
+    if (length(list(...)) > 0) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:276]:
```suggestion
    if (...length() > 0) {
```

File: R/magpie-class.R [hunk 2]
```diff
@@ -302,39 +305,27 @@ setMethod("[", #nolint
...
     if (!missing(k)) k <- .dimextract(x, k, 3, pmatch = pmatch, invert = invert)
-
-    if (!missing(i) && missing(j) && !missing(k) && .isFALSE(k) &&
-      .isFALSE(drop) && .isFALSE(pmatch) && .isFALSE(invert)) {
-      # there is a weird case in which k is actually missing but is getting the
-      # value of the next argument in line (drop)
-      # this one is catched via .isFALSE(k)
-      # in addition non-default settings for drop, pmatch and invert indicate that object should still be
-      # returned as magpie object (and not as numeric as the following line will do)
+    if (length(sys.call(-1)) == 3) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:308]: Please add a comment what is happening here and why

File: R/magpie-class.R [hunk 3]
```diff
@@ -302,39 +305,27 @@ setMethod("[", #nolint
...
   }
 )
 
 #' @exportMethod [<-
-setMethod("[<-", #nolint
+setMethod("[<-", # nolint
   signature(x = "magpie"),
-  function(x, i, j, k, value, pmatch = FALSE, dim = NULL) {
+  function(x, i, j, k, ..., pmatch = FALSE, dim = NULL, value) {
+    if (length(list(...)) > 0) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:321]:
```suggestion
    if (...length() > 0) {
```

File: R/magpie-class.R [hunk 4]
```diff
@@ -365,26 +356,25 @@ setMethod("[<-", #nolint
...
-      } else if (length(value) != length(x@.Data[i, j, k]) & length(value) != 1) {
-        # dangerous writing of value as order might be wrong!
-        stop("Replacement does not work! Different replacement length!")
-      } else if (length(value) != 1 && !is.null(getOption("magclass.verbosity")) &&
-                 getOption("magclass.verbosity") > 1) {
-        message("NOTE ([<-): Dangerous replacement! As replacement value is not",
-          " an MAgPIE object name checking is deactivated!\n")
-      }
-      x@.Data[i, j, k] <- value
+    if (length(sys.call(which = -1)) == 4) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:359]: Can you comment what is happening here and why?

File: R/magpie-class.R [hunk 5]
```diff
@@ -365,26 +356,25 @@ setMethod("[<-", #nolint
...
+      x@.Data[i] <- value
       return(x)
     }
+    if (is.magpie(value)) {
+      if (missing(i)) ii <- seq_len(dim(x)[1]) else ii <- i
+      if (missing(j)) jj <- seq_len(dim(x)[2]) else jj <- j
+      if (missing(k)) kk <- seq_len(dim(x)[3]) else kk <- k
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:366]: I personally find the following a bit easier to read, but feel free to decline this.
```suggestion
      ii <- if (missing(i)) seq_len(dim(x)[1]) else i
      jj <- if (missing(j)) seq_len(dim(x)[2]) else j
      kk <- if (missing(k)) seq_len(dim(x)[3]) else k
```

File: R/magpie-class.R [hunk 6]
```diff
@@ -365,26 +356,25 @@ setMethod("[<-", #nolint
...
+      value <- magpie_expand(value, x[ii, jj, kk])
+    } else if (length(value) != length(x@.Data[i, j, k]) & length(value) != 1) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:368]: Did the linter not flag this? I thought `&` is always a linter warning in `if` statements
```suggestion
    } else if (length(value) != length(x@.Data[i, j, k]) && length(value) != 1) {
```

File: R/magpie-class.R [hunk 7]
```diff
@@ -365,26 +356,25 @@ setMethod("[<-", #nolint
...
+      # dangerous writing of value as order might be wrong!
+      stop("Replacement does not work! Different replacement length!")
+    } else if (length(value) != 1 && !is.null(getOption("magclass.verbosity")) &&
+               getOption("magclass.verbosity") > 1) {
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:372]:
```suggestion
    } else if (length(value) != 1 && getOption("magclass.verbosity", default = 0) > 1) {
```

File: R/magpie-class.R [hunk 8]
```diff
@@ -365,26 +356,25 @@ setMethod("[<-", #nolint
...
+      message("NOTE ([<-): Dangerous replacement! As replacement value is not",
+        " an MAgPIE object name checking is deactivated!\n")
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:374]: If this really is dangerous, did you consider throwing a warning here instead?

**Review by pfuehrlich-pik — CHANGES_REQUESTED** [review]

File: R/magpie-class.R (hunk 2 above)
Comment (tscheypidi) [reply R/magpie-class.R:310]: done

File: R/magpie-class.R (hunk 5 above)
Comment (tscheypidi) [reply R/magpie-class.R:370]: I would like to keep the changes to a minimum for now

File: R/magpie-class.R [hunk 9]
```diff
@@ -356,6 +358,8 @@ setMethod("[<-", # nolint
       if (is.factor(k)) k <- as.character(k)
       if (is.character(k) || is.list(k)) k <- .dimextract(x, k, 3, pmatch = pmatch)
     }
+    # if [] contains only 1 argument, the number of elements in sys.call(-1)
+    # is 4. This indicates that it is a vectorised call of the form x[i]
```
Comment (pfuehrlich-pik) [inline R/magpie-class.R:362]:
```suggestion
    # is 4. This indicates that it is a vectorised call of the form x[i] <- value
```

**Review by pfuehrlich-pik — APPROVED** [review]

**Pull request merged**